main.go

```diff
@@ -2,208 +2,243 @@ package main
-	e.usage.Describe(ch)
-	e.bytes.Describe(ch)
-	e.removals.Describe(ch)
+	ch <- e.up
+	ch <- e.uptime
+	ch <- e.version
+	ch <- e.bytesRead
+	ch <- e.bytesWritten
+	ch <- e.connections
+	ch <- e.connectionsTotal
+	ch <- e.currentBytes
+	ch <- e.limitBytes
+	ch <- e.commands
+	ch <- e.items
+	ch <- e.itemsTotal
+	ch <- e.evictions
+	ch <- e.reclaimed
 }
 
-// Collect fetches the statistics from the configured memcache servers, and
+// Collect fetches the statistics from the configured memcached servers, and
 // delivers them as prometheus metrics. It implements prometheus.Collector.
 func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
-	// prevent concurrent metric collections
-	e.mutex.Lock()
-	defer e.mutex.Unlock()
-
-	e.cache.Reset()
-	e.usage.Reset()
-	e.bytes.Reset()
-	e.removals.Reset()
-
 	stats, err := e.mc.Stats()
-
 	if err != nil {
-		e.up.Set(0)
-		ch <- e.up
-		glog.Infof("Failed to collect stats from memcache: %s", err)
+		ch <- prometheus.MustNewConstMetric(e.up, prometheus.GaugeValue, 0)
+		log.Errorf("Failed to collect stats from memcached: %s", err)
 		return
 	}
+	ch <- prometheus.MustNewConstMetric(e.up, prometheus.GaugeValue, 1)
 
-	e.up.Set(1)
-	for server, _ := range stats {
-		m, err := strconv.ParseUint(stats[server]["uptime"], 10, 64)
-		if err != nil {
-			e.uptime.Set(0)
-		} else {
-			e.uptime.Set(float64(m))
-		}
-
-		for _, op := range cacheOperations {
-			for _, st := range cacheStatuses {
-				m, err := strconv.ParseUint(stats[server][fmt.Sprintf("%s_%s", op, st)], 10, 64)
-				if err != nil {
-					e.cache.WithLabelValues(op, st).Set(0)
-				} else {
-					e.cache.WithLabelValues(op, st).Set(float64(m))
-				}
-			}
-		}
+	for _, s := range stats {
+		ch <- prometheus.MustNewConstMetric(e.uptime, prometheus.CounterValue, parse(s, "uptime"))
+		ch <- prometheus.MustNewConstMetric(e.version, prometheus.GaugeValue, 1, s["version"])
 
-		for _, t := range usageTimes {
-			for _, r := range usageResources {
-				m, err := strconv.ParseUint(stats[server][fmt.Sprintf("%s_%s", t, r)], 10, 64)
-				if err != nil {
-					e.usage.WithLabelValues(t, r).Set(0)
-				} else {
-					e.usage.WithLabelValues(t, r).Set(float64(m))
-				}
-			}
+		for _, op := range []string{"get", "delete", "incr", "decr", "cas", "touch"} {
+			ch <- prometheus.MustNewConstMetric(e.commands, prometheus.CounterValue, parse(s, op+"_hits"), op, "hit")
+			ch <- prometheus.MustNewConstMetric(e.commands, prometheus.CounterValue, parse(s, op+"_misses"), op, "miss")
 		}
+		ch <- prometheus.MustNewConstMetric(e.commands, prometheus.CounterValue, parse(s, "cas_badval"), "cas", "badval")
+		ch <- prometheus.MustNewConstMetric(e.commands, prometheus.CounterValue, parse(s, "cmd_flush"), "flush", "hit")
 
-		for _, dir := range bytesDirections {
-			m, err := strconv.ParseUint(stats[server][fmt.Sprintf("bytes_%s", dir)], 10, 64)
-			if err != nil {
-				e.bytes.WithLabelValues(dir).Set(0)
+		// memcached includes cas operations again in cmd_set.
```

Comment:
badval is not an error condition, just a client losing in an update race. I think it's correct like this – this is just the only case where a third outcome is possible.